torch_disc/torch_disc/csrc/disc_jit.cpp

```diff
@@ -0,0 +1,212 @@
+      group_out->setType(out->type());
+    }
+    node->destroy();
+    CastToTensorInputs(graph, group, node_merged);
+  }
+  torch::jit::EliminateDeadCode(graph);
+}
+
+std::string DiscCMD(const std::string& mlir_fname,
+                    const std::string& out_fname) {
+  std::stringstream ss;
+  std::string logf = mlir_fname + ".log";
+  ss << "./disc_compiler_main " << mlir_fname << " " << out_fname << " > "
+     << logf << " 2>&1 ";
+  return ss.str();
+}
+
+std::string GetTempDirectory(std::string dir) {
+  auto tid = std::this_thread::get_id();
+  uint64_t pid = getpid();
+  uint64_t us =
+      std::chrono::duration_cast<std::chrono::microseconds>(
+          std::chrono::high_resolution_clock::now().time_since_epoch())
+          .count();
+  std::stringstream ss;
+  ss << dir << "/" << tid << "-" << pid << "-" << us;
+  TORCH_CHECK(!mkdir(ss.str().c_str(), 0755),
+              "unable to create dir: " + ss.str());
+  return ss.str();
+}
+
+std::string MhloConversaion(const std::shared_ptr<Graph>& graph) {
+  std::string parsable_mlir;
+  std::string pretty_mlir;
+  std::string input_dev_str;
+  std::string output_dev_str;
+  std::tie(parsable_mlir, pretty_mlir, input_dev_str, output_dev_str) =
+      torch::blade::ConvertTorchScriptToMhlo(graph);
+  TORCH_CHECK(!parsable_mlir.empty(), "mhlo conversaion failed!");
+
+  std::string dir = GetTempDirectory("/tmp");
+  std::string in_fname = dir + "/disc.mlir";
+  std::ofstream outfile(in_fname);
+  outfile << parsable_mlir << std::endl;
+  outfile.flush();
+  return in_fname;
+}
+
+void CallDiscCompiler(const std::string& mlir_fname) {
+  std::string out_fname = mlir_fname + ".out";
+  std::string cmd = DiscCMD(mlir_fname, out_fname);
+  TORCH_CHECK(std::system(cmd.c_str()) == 0,
+              "disc compile failed with cmd: " + cmd);
+}
+
+void DiscCompilation(const std::shared_ptr<Graph>& graph) {
+  for (auto node : graph->nodes()) {
+    if (node->kind() == prim::FusionGroup) {
+      auto sub_graph = node->g(attr::Subgraph);
+      std::string fname = MhloConversaion(sub_graph);
+      CallDiscCompiler(fname);
+    }
+  }
+}
+
+void InferShapes(const std::shared_ptr<Graph>& graph,
+                 c10::ArrayRef<torch::lazy::BackendDataPtr> arguments) {
+  for (size_t i = 0; i < graph->inputs().size(); ++i) {
+    auto argument = arguments[i];
+    auto input = graph->inputs()[i];
+    const auto ts_data = std::static_pointer_cast<TSData>(argument);
+
+    if (ts_data->HasValue()) {
+      input->setType(c10::TensorType::create(ts_data->data()));
+    }
+  }
+  torch::jit::PropagateInputShapes(graph);
+}
+
+void DiscJIT(torch::lazy::TSComputation& computation,
+             c10::ArrayRef<torch::lazy::BackendDataPtr> arguments) {
+  auto graph = computation.graph()->copy();
+  // 1. clustering and group DISC nodes
+  // 2. conversion from torchscript Graph to mhlo Dialect on DISC nodes
+  // 2. register DISC engine
+  InferShapes(graph, arguments);
+  FusionDiscNodes(graph);
```

Comment:
We'd better avoid to use  "fusion" here, which can easily be misunderstood

Reply:
Good point. Actually, this function just **CLUSTER** instead of **FUSION**, maybe better to ClusterDiscNodes ?